lib/workspace.py

```diff
@@ -143,7 +143,8 @@ def update(self):
 
                 # 4. If the repo has a selected version, fail if that version
                 # does not include this tuple's hash
-                if not repo.is_ancestor(commit, selected_commit):
+                if not (repo.is_ancestor(commit, selected_commit)
```

Comment:
I had to think through the algorithm again to wrap my head around why the code was originally written as it was. So this is an interesting question that unfortunately rests on the /intention/ of the repo authors.

If the author of `bar:2018` explicitly specified `xyz:2010` despite newer versions existing (which is the only way that `foo:2017` could have a newer version: `xyz:2011`), is it safe to override that requirement with the newer version?

The commit-time based ordering of the repos is intentional and this somewhat defeats that. I'm trying to think through a case where this causes problems; if `xyz` has a set of dependencies that need to be processed, we could end up in a weird state... Curious to hear @jackkoenig and @terpstra thoughts on this one.

Reply:
Alright I discussed with @terpstra and the algorithm *should* support this test case, the implementation of the algorithm is slightly wrong. I haven't read the code, but in the example:
```
top:2019
├─ bar:2018
│  └─ xyz:2010
└─ foo:2017
   └─ xyz:2011
```
We visit `top` (pushing `bar` and `foo` into the priority queue), then visit `bar` and push `xyz:2010` to the priority queue, then we visit `foo` and push `xyz:2011` to the priority queue, then `xyz:2011` is newer than `xyz:2010` so we visit `xyz:2011`.

This fix is wrong, @aaronjanse is investigating

Reply:
The issue is that in order to know the timestamp of a commit, we have to clone the repo.

We're cloning `xyz` from `bar` to figure out whether we should explore it before or after `foo`.

```bash
# FIXME: This should clone to a temporary area. If this were
# fixed we could get rid of the source_map dictionary hack
```
If we clone to a temporary directory, as suggested in the existing code comment, we would have seen that `xyz:2010` is older _before_ cloning it to the workspace.

Reply:
Sorry for the drive-by commenting, since I have no context on this codebase, but you can't make use of git commit timestamps for ordering since interactive rebasing allows one to create a git commit graph that does not match the chronological ordering. Not to mention that the `--date` option to `git commit` allows one to override the date of a commit.

This is something I've noticed that even GitHub gets wrong, since I'll sometimes reorder my commits before submitting a pull request, and GitHub will display them in chronological order, not commit history order. This is most noticeable when the tip of the branch has a commit date that is older than another commit in the history, and the CI status icon appears on a commit that is not at the bottom of the page.